feat(backend): start_and_restart_docker_container_even_when_it_is_already_registered_in_Docker

alex1307 opened the pull request:
Description

If container for a given tari image exists in Docker then it can't be started or restarted.

Motivation and Context

Launchpad should have full controll of tari containers and images.
Containers should be destroyed after shutdown, but it may happen to have abondened container.
On start we should check whether there is a docker container and destroy it.
Then we should create fresh new image and start it.

How Has This Been Tested?

Manually: Run Launchpad app, started and stopped containers.
Started containers from Launchpad and stopped manually on Docker.
Started the containers from Launchpad again.

alex1307 committed:
feat(backend): start/restart docker container even when it is already registered in Docker
Description

If container for a given tari image exists in Docker then it can't be started or restarted.

Motivation and Context

Launchpad should have full controll of tari containers and images.
Containers should be destroyed after shutdown, but it may happen to have abondened container.
On start we should check whether there is a docker container and destroy it.
Then we should create fresh new image and start it.

How Has This Been Tested?

Manually: Run Launchpad app, started and stopped containers.
Started containers from Launchpad and stopped manually on Docker.
Started the containers from Launchpad again.

CjS77 commented on applications/launchpad/backend/src/docker/mod.rs:
```diff
@@ -57,3 +62,90 @@ lazy_static! {
+    format!("{}_{}_volume", tari_workspace, tari_network.lower_case())
+}
+
+pub fn network_name(tari_workspace: &str) -> String {
```
Confusingly named function.

This doesn't return the network name, but the key for this workspace's entry in a hashmap.

```suggestion
pub fn network_key(tari_workspace: &str) -> String {
```
```suggestion
pub fn network_name(tari_workspace: &str) -> String {
```

CjS77 replied:
1. Can this be a private function?
2. Can `tari_blockchain_volume_name` be private?

CjS77 commented on applications/launchpad/backend/src/docker/mod.rs:
```diff
@@ -57,3 +62,90 @@ lazy_static! {
+    format!("{}_{}_volume", tari_workspace, tari_network.lower_case())
+}
+
+pub fn network_name(tari_workspace: &str) -> String {
```
Confusingly named function.

This doesn't return the network name, but the key for this workspace's entry in a hashmap.

```suggestion
pub fn network_key(tari_workspace: &str) -> String {
```
```suggestion
pub fn network_name(tari_workspace: &str) -> String {
```

CjS77 commented on applications/launchpad/backend/src/docker/mod.rs:
```diff
@@ -57,3 +62,90 @@ lazy_static! {
+        .await?;
+
+    for c_id in containers.iter() {
+        let id = c_id.clone().id.unwrap();
```
1. Is this `clone` necessary? Looks like you can avoid it
2. Is the unwrap infallible?

CjS77 commented on applications/launchpad/backend/src/docker/settings.rs:
```diff
@@ -42,13 +42,13 @@ http://monero-stagenet.exan.tech:38081,\
 http://singapore.node.xmr.pm:38081";
 
-#[derive(Default, Debug, Serialize, Deserialize)]
+#[derive(Default, Debug, Serialize, Deserialize, Clone)]
```
If you don't actually clone these structs, don't bloat the codebase with the derived functions.

alex1307 replied:
pure java style :). Fixed

CjS77 commented on applications/launchpad/backend/src/docker/mod.rs:
```diff
@@ -57,3 +62,90 @@ lazy_static! {
+    image: ImageType,
+    fully_qualified_image_name: String,
+    tari_workspace: &str,
+    config: LaunchpadConfig,
```
Pretty sure you can get away with a reference here

alex1307 replied:
Changes are applied.

CjS77 reviewed (commented):
Some minor issues. Generall y looks good.

alex1307 committed:
feat(backend): 239_restart_docker_containers
Description

Fixing code quality issues:
 Applying cargo fmt --all
 Delete commented code.
 Delete Clone, make functions private.

Motivation and Context
Clean code quality requirement.

How Has This Been Tested?
The change is manually tested.

alex1307 committed:
feat(backend): restart docker containers
Description

Fixing code quality issues:
Fixing potential source of panicking a thread:
	Replaced  c_id.id.clone()unwrap() with if let Some(id) = c_id.id.clone()

Motivation and Context
Clean code quality requirement.

How Has This Been Tested?
The change is tested manually.

CjS77 reviewed (dismissed):
LGTM

alex1307 committed:
Merge branch 'launchpad_such_wow' into 239_restart_docker_containers

CjS77 reviewed (approved):
+1
